### PR TITLE
fix sample for Sycl 2020 not supporting mutable

### PR DESCRIPTION
**Pull request by terdner**
# Description

Sycl 2020 doesn't support mutable and the sample had several cases that needed to be changed. 

Fixes # (issue) 
https://jira.devtools.intel.com/browse/ONSAM-1351 

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Sample Migration (Moving sample from old repository after completing checklist established)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode

# Checklist for Moving samples:
Links and Details can be found in the samples WG Teams Files. 

- [ ] Review sample design with domain reviewers https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts 
- [ ] Implement coding guidelines and ensure code quality.
- [ ] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com
- [ ] Adhere to readme template 
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Enforce format via clang-format config file
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth or Joe Oster. (GitHub User: JoeOster)
- [ ] Tested using Dev Cloud when applicable
- [ ] Implement fixes for ONSAM Jiras
- [ ] If you have new dependencies/binaries, inform Samples Project Manager Swapna R Dontharaju (@srdontha) or @JoeOster

### CODE REVIEW

File: DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp [hunk 1]
```diff
@@ -163,32 +163,36 @@ struct slice_area {
...
-struct accessor_traits<RawArrayValueType*>
+template <typename _RawArrayValueType>
+struct accessor_traits_impl<_RawArrayValueType*>
 {
-    using value_type = RawArrayValueType;
+    using value_type = _RawArrayValueType;
 };
 
+template <typename _Unknown>
+using accessor_traits = accessor_traits_impl<typename std::decay<_Unknown>::type>;
```
Comment (capatober) [inline DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp:193]: Feel free to get rid of the underscore before template arguments (here and above) if it looks ugly

File: DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp [hunk 2]
```diff
@@ -163,32 +163,36 @@ struct slice_area {
...
+
+
```
Comment (capatober) [inline DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp:195]: This empty line seems extra

File: DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp [hunk 3]
```diff
@@ -748,8 +751,8 @@ int main(int argc, char** argv) {
 
   // Use the DPC++ library call to reduce the array using plus
   buffer<float> calc_values(results_per_rank, num_step_per_rank);
-  auto calc_begin2 = dpstd::begin(calc_values);
-  auto calc_end2 = dpstd::end(calc_values);
+  auto calc_begin2 = oneapi::dpl::begin(calc_values);
+  auto calc_end2 =  oneapi::dpl::end(calc_values);
 
   local_sum =
       std::reduce(policy, calc_begin2, calc_end2, 0.0f, std::plus<float>());
```
Comment (capatober) [inline DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp:758]: `std::reduce` is defined in `<numeric>`. Could you please include this header?
By the way, `std::reduce` is blocking, you don't need an explicit `wait()` below.

File: DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp [hunk 4]
```diff
@@ -316,7 +320,6 @@ float calc_pi_dpstd_native3(size_t num_steps, int groups, Policy&& policy) {
   auto calc_end = oneapi::dpl::end(buf);
```
Comment (capatober) [inline DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp:320]: I see `calc_begin`, `calc_end` are not used in this function. It also correct for the `calc_pi_dpstd_native4` function

File: DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp (hunk 1 above)
Comment (terdner) [reply DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp:193]: will fix

File: DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp (hunk 4 above)
Comment (terdner) [reply DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp:320]: will fix

File: DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp (hunk 2 above)
Comment (terdner) [reply DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp:195]: will fix

**Comment by capatober** [comment]
Overall, I'm ok with changes

File: DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp (hunk 3 above)
Comment (terdner) [reply DirectProgramming/DPC++/ParallelPatterns/dpc_reduce/src/main.cpp:750]: fixed in my new push

**Review by capatober — APPROVED** [review]

**Review by JoeOster — APPROVED** [review]

**Pull request merged**